core/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/PlainHighlighter.java

```diff
@@ -104,9 +105,14 @@ public HighlightField highlight(HighlighterContext highlighterContext) {
 
         try {
             textsToHighlight = HighlightUtils.loadFieldValues(field, mapper, context, hitContext);
-
             for (Object textToHighlight : textsToHighlight) {
-                String text = textToHighlight.toString();
+                String text;
+                if (textToHighlight instanceof BytesRef) {
+                    // keywords are internally stored as utf8 bytes
+                    text = ((BytesRef) textToHighlight).utf8ToString();
+                } else {
+                    text = textToHighlight.toString();
+                }
```

Comment:
would it work if we called MappedFieldType.valueForDisplay instead? I'm concerned some fields are stored as bytesrefs too but do not represent utf8 strings, like ip addresses